pkg/kubelet/kuberuntime/logs/logs_test.go

```diff
@@ -210,6 +213,144 @@ func TestReadLogs(t *testing.T) {
 	}
 }
 
+type syncBuffer struct {
+	b bytes.Buffer
+	sync.Mutex
+}
+
+func (b *syncBuffer) Read(p []byte) (n int, err error) {
+	b.Lock()
+	defer b.Unlock()
+	return b.b.Read(p)
+}
+func (b *syncBuffer) Write(p []byte) (n int, err error) {
+	b.Lock()
+	defer b.Unlock()
+	return b.b.Write(p)
+}
+
+func (b *syncBuffer) String() string {
+	b.Lock()
+	defer b.Unlock()
+	return b.b.String()
+}
+
+func (b *syncBuffer) Reset() {
+	b.Lock()
+	defer b.Unlock()
+	b.b.Reset()
+}
+
+func TestReadRotatedLog(t *testing.T) {
+	tmpDir := t.TempDir()
+	file, err := os.CreateTemp(tmpDir, "logfile")
+	if err != nil {
+		assert.NoErrorf(t, err, "unable to create temp file")
+	}
+
+	stdoutBuf := &syncBuffer{}
+	stderrBuf := &syncBuffer{}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	// Start to follow the container's log.
+	go func(ctx context.Context) {
+		containerID := "fake-container-id"
+		fakeRuntimeService := &apitesting.FakeRuntimeService{
+			Containers: map[string]*apitesting.FakeContainer{
+				containerID: {
+					ContainerStatus: runtimeapi.ContainerStatus{
+						State: runtimeapi.ContainerState_CONTAINER_RUNNING,
+					},
+				},
+			},
+		}
+		podLogOptions := v1.PodLogOptions{
+			Follow: true,
+		}
+		opts := NewLogOptions(&podLogOptions, time.Now())
+		ReadLogs(ctx, file.Name(), containerID, opts, fakeRuntimeService, stdoutBuf, stderrBuf)
+	}(ctx)
+
+	// log in stdout
+	expectedStdout := "line0\nline2\nline4\nline6\nline8\n"
+	// log in stderr
+	expectedStderr := "line1\nline3\nline5\nline7\nline9\n"
+
+	dir := filepath.Dir(file.Name())
+	baseName := filepath.Base(file.Name())
+
+	end := make(chan struct{})
+	go func() {
+		timeout := time.NewTimer(30 * time.Second)
+		ticker := time.NewTicker(10 * time.Millisecond)
+		defer func() {
+			if !timeout.Stop() {
+				<-timeout.C
+			}
+			ticker.Stop()
+		}()
+
+		for {
+			select {
+			// When podLogOptions.Follow is true, the function ReadLogs will not end.
+			// So we should check whether stdoutBuf && stderrBuf match the expects periodically.
```

Comment:
Would it end if we delete the file? If so, we can test it without timer, timeout and context, I hope.